geopandas/tools/sjoin.py

```diff
@@ -27,131 +31,76 @@ def sjoin(left_df, right_df, how='inner', op='intersects',
-            or any(right_df.columns.isin([index_left, index_right]))):
-        raise ValueError("'{0}' and '{1}' cannot be names in the frames being"
-                         " joined".format(index_left, index_right))
-
-    # the rtree spatial index only allows limited (numeric) index types, but an
-    # index in geopandas may be any arbitrary dtype. so reset both indices now
-    # and store references to the original indices, to be reaffixed later.
-    # GH 352
-    left_df.index.name = index_left
-    right_df.index.name = index_right
-    left_df = left_df.reset_index()
-    right_df = right_df.reset_index()
-
-    if op == "within":
-        # within implemented as the inverse of contains; swap names
-        left_df, right_df = right_df, left_df
-
-    # insert the bounds in the rtree spatial index
-    right_df_bounds = right_df.geometry.apply(lambda x: x.bounds)
-    stream = ((i, b, None) for i, b in enumerate(right_df_bounds))
-    tree_idx = rtree.index.Index(stream)
-
-    idxmatch = (left_df.geometry.apply(lambda x: x.bounds)
-                .apply(lambda x: list(tree_idx.intersection(x))))
-    idxmatch = idxmatch[idxmatch.apply(len) > 0]
-
-    if idxmatch.shape[0] > 0:
-        # if output from join has overlapping geometries
-        r_idx = np.concatenate(idxmatch.values)
-        l_idx = np.concatenate([[i] * len(v) for i, v in idxmatch.iteritems()])
-
-        # Vectorize predicate operations
-        def find_intersects(a1, a2):
-            return a1.intersects(a2)
-
-        def find_contains(a1, a2):
-            return a1.contains(a2)
-
-        predicate_d = {'intersects': find_intersects,
-                       'contains': find_contains,
-                       'within': find_contains}
-
-        check_predicates = np.vectorize(predicate_d[op])
-
-        result = (
-                  pd.DataFrame(
-                      np.column_stack(
-                          [l_idx,
-                           r_idx,
-                           check_predicates(
-                               left_df.geometry
-                               .apply(lambda x: prepared.prep(x))[l_idx],
-                               right_df[right_df.geometry.name]._geometry_array[r_idx])
-                           ]))
-                   )
-
-        result.columns = ['_key_left', '_key_right', 'match_bool']
-        result = (
-                  pd.DataFrame(result[result['match_bool']==1])
-                  .drop('match_bool', axis=1)
-                  )
-
+    indices = cysjoin(left_df.geometry._geometry_array.data,
+                      right_df.geometry._geometry_array.data,
+                      op)
+    n = len(indices)
+    left_indices = indices[:, 0]
+    right_indices = indices[:, 1]
+
+    if how == 'left':
+        missing = pd.Index(np.arange(len(left_df))).difference(pd.Index(left_indices))
+        if len(missing):
+            left_indices = np.concatenate([left_indices, missing])
```

Comment:
The reason that I wonder it, is because (IIUC) that now entries in the left without a match will all end at the bottom (in case of a left/inner join), while it could make sense to keep the original ordering of the left frame ?